### ISSUE TITLE
Can we have eqn: for [induction] too?

### ISSUE THREAD

**Issue by bcpierce00** [coq/coq#8466]
I've just been updating SF to add [eqn:] almost everywhere.  This makes an amazing difference in how easy it is to walk through proofs and explain them: It makes it immediately clear which case of a big case analysis you are in, and it automatically prevents some confusing reuse of names (e.g., if you destruct a number n, you don't now have n as the name of the predecessor of the old n) if you leave off the [as] clause.  Excellent improvement!

So now I'm sad that I can't (at least in 8.8.1) do the same for the [induction] tactic.  Would it be hard to add?

**Comment by SkySkimmer** [comment]
Does it make sense to add to induction? I can't see how it would work.
ie what should this do:
~~~coq
Definition deep_id (n:nat) : nat.
Proof.
  induction n eqn:H.
  (* H : ??? *)
~~~

**Comment by bcpierce00** [comment]
Does adding 
     `H : n=0`
to the context at this point not make sense?

**Comment by SkySkimmer** [comment]
I don't think so, it will be reached in recursive calls as well.

**Comment by Zimmi48** [comment]
@bcpierce00 :+1:
A bit off-topic, but do you think that we should make `destruct` behave like it has a `eqn:?` clause by default (unless it has an `as` clause that reuses the name that is destructed)? That would be a compatibility breaking change though, especially for people relying on auto-generated names.

**Comment by gares** [comment]
SSR has this feature as in `elim H : x` but it is mostly a debug feature since it makes the induction hypotheses unreachable most of the times.

**Comment by JasonGross** [comment]
@bcpierce00 it sounds like you want something like
```coq
Ltac myinduction H :=
  let H' := fresh H in
  pose H as H';
  induction H;
  first [ rename H' into H | rename H' into H, H' into H ].
```
?

> A bit off-topic, but do you think that we should make `destruct` behave like it has a `eqn:?` clause by default

I oppose this change unless someone (not me) writes a script that automatically updates all existing code to be compatible with this change.  (I would not oppose having a setting that changes the default, though I'd prefer if the feature for "locally set this option in this bit of tactic code" lands first.)

**Comment by Zimmi48** [comment]
>I oppose this change unless someone (not me) writes a script that automatically updates all existing code to be compatible with this change.

Yeah, I realize that changing the default would be a major pain. But suppose this was not a concern (very theoretical), what would you feel about such a default?

**Comment by bcpierce00** [comment]
> @bcpierce00 <https://github.com/bcpierce00> it sounds like you want something like
> Ltac myinduction H :=
>   let H' := fresh H in
>   pose H as H';
>   induction H;
>   first [ rename H' into H | rename H' into H, H' into H ].
> ?
> 
Yep!

(And I want it to have an optional eqn:H’ annotation, and an optional “as” annotation, and it I want it to be called “induction”. :-)
> A bit off-topic, but do you think that we should make destruct behave like it has a eqn:? clause by default
> 
> I oppose this change unless someone (not me) writes a script that automatically updates all existing code to be compatible with this change. (I would not oppose having a setting that changes the default, though I'd prefer if the feature for "locally set this option in this bit of tactic code" lands first.)
> 
I agree that changing the default for everybody would be a huge hassle, but it might indeed be nice to be able to do so locally for teaching materials.

    - B

**Comment by JasonGross** [comment]
> But suppose this was not a concern (very theoretical), what would you feel about such a default?

The main thing I'd care about is "how many characters do I have to type on average", and I think I use the equation version a bit more, so I feel slightly positive about it.  But, really, don't optimize the defaults for people like me who mainly write tactic scripts.
